codalab/client/json_api_client.py

```diff
@@ -0,0 +1,299 @@
+    elif isinstance(data, dict):
+        result = unpack_object(document['data'])
+    else:
+        result = {}
+
+    # Include meta
+    # (Warning: this may overwrite meta present at the resource object level.)
+    if 'meta' in document:
+        result['meta'] = document['meta']
+
+    return result
+
+
+def pack_document(objects, type_):
+    """
+    Pack resource object(s) into a JSON API request document.
+    References to relationships should be indicated by a placeholder
+    JsonApiRelationship object. For example:
+
+     >> pack_document({
+            'owner': JsonApiRelationship('users', '345'),
+            'id': '123',
+            'name': 'hello'
+        }, 'bundles')
+
+     => {
+            'data': {
+                'id': '123',
+                'type': 'bundles',
+                'attributes': {
+                    'name': 'hello'
+                },
+                'relationships': {
+                    'owner': {
+                        'data': {
+                            'id': '345',
+                            'type': 'users'
+                        }
+                    }
+                },
+            }
+        }
+
+    :param objects: a dict or list of dicts representing resources
+    :param type_: resource type
+    :return:
+    """
+    def pack_object(obj):
+        packed_obj = {'type': type_}
+        attributes = {}
+        relationships = {}
+        for key, value in obj.iteritems():
+            if isinstance(value, JsonApiRelationship):
+                relationships[key] = {
+                    'data': {
+                        'id': value.id_,
+                        'type': value.type_,
+                    }
+                }
+            elif key == 'id':
+                packed_obj['id'] = value
+            else:
+                attributes[key] = value
+
+        if attributes:
+            packed_obj['attributes'] = attributes
+
+        if relationships:
+            packed_obj['relationships'] = relationships
+
+        return packed_obj
+
+    if isinstance(objects, list):
+        packed_objects = [pack_object(obj) for obj in objects]
+    else:
+        packed_objects = pack_object(objects)
+
+    return {
+        'data': packed_objects
+    }
+
+
+class JsonApiRelationship(object):
+    """
+    Placeholder for a relationship to another resource.
+    Used to build requests to create or update a resource.
+
+    For example, to update the owner of a bundle:
+        client.update('bundles', uuid, data={
+            'owner': JsonApiRelationship('users', owner_id)
+        })
+    """
+    def __init__(self, type_, id_):
+        self.type_ = type_
+        self.id_ = id_
+
+
+class JsonApiClient(RestClient, BundleClient):
+    """
+    Simple JSON API client.
+    """
+    def __init__(self, address, auth_handler, authenticate):
+        self.address = address
+        self.auth_handler = auth_handler
```

Comment:
These are used by the CodalabManager to authenticate, I'll add comments to clarify.